### PR TITLE
feat: Add `cache-key-prefix` option

### PR DESCRIPTION
**Pull request by erezrokah**
**Description:**

Adds a cache key prefix option so it's possible to use different cache keys for different jobs

**Related issue:**

Fixes https://github.com/actions/setup-go/issues/358

**Check list:**
- [x] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by erezrokah** [comment]
Not sure why the windows tests are failing, so any help will be appreciated

**Comment by erezrokah** [comment]
Working examples in https://github.com/cloudquery/cloudquery/actions/runs/4660341686/jobs/8248290837#step:3:26 and https://github.com/cloudquery/cloudquery/actions/runs/4660341686/jobs/8248291007#step:4:23 via https://github.com/cloudquery/cloudquery/pull/9812

**Comment by rokerzfirst101** [comment]
> Not sure why the windows tests are failing, so any help will be appreciated

I recently forked the repo, made the same changes, and got the same error for the linter workflow. Somehow, rerunning the job with debug logging made it pass.

**Comment by erezrokah** [comment]
Hello lovely maintainers 👋 Anything we can do to move this forward (or close it so we'll if know if we need to use a different solution)?

**Comment by candiduslynx** [comment]
> Hello lovely maintainers 👋 Anything we can do to move this forward (or close it so we'll if know if we need to use a different solution)?

@dsame @MaksimZhukov @IvanZosimov @marko-zivic-93

**Comment by marko-zivic-93** [comment]
Hello everyone,

Sorry for the late response, first of all 🙂 
We have started working on this feature, but right now we have shifted our focus a bit from it. This should be handled in the upcoming quarter. Thank you for your patience 🙂

**Comment by dsame** [comment]
Hello @erezrokah , can you please review and apply the changes i requested in order we can approve the PR merge?

**Comment by candiduslynx** [comment]
> Hello @erezrokah , can you please review and apply the changes i requested in order we can approve the PR merge?

@dsame did you submit the review, though? I can't see any comments ATM

**Comment by erezrokah** [comment]
> Hello @erezrokah , can you please review and apply the changes i requested in order we can approve the PR merge?

Thanks for the review @dsame, but I'm not seeing it. Can you confirm you submitted it?

**Comment by dsame** [comment]
@erezrokah please take a look at this one https://github.com/actions/setup-go/pull/366/files/acb2fd804097c1dffc2eb1075b3696da4b4bb27c#r1158511364 ?

**Comment by candiduslynx** [comment]
> @erezrokah please take a look at this one https://github.com/actions/setup-go/pull/366/files/acb2fd804097c1dffc2eb1075b3696da4b4bb27c#r1158511364 ?

@dsame 
You definitely didn't publish the review (kindly check that you oush `submit review` in the review tab)

**Comment by erezrokah** [comment]
> @erezrokah please take a look at this one https://github.com/actions/setup-go/pull/366/files/acb2fd804097c1dffc2eb1075b3696da4b4bb27c#r1158511364 ?

I'm not seeing anything there. Could it be that your review comments are still pending and you need to submit them like @candiduslynx mentioned? See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request#submitting-your-review

**Comment by dsame** [comment]
@candiduslynx @erezrokah thanks for you note, i missed to click "submit" indeed. No the review is published

File: src/cache-restore.ts
Comment (candiduslynx) [inline src/cache-restore.ts:33]: from @dsame [comment](https://github.com/actions/setup-go/pull/366/files/acb2fd804097c1dffc2eb1075b3696da4b4bb27c#r1212666065)
```suggestion
  const cacheKeyPrefix = core.getInput('cache-key-prefix') || 'setup-go'
  const primaryKey = `${cacheKeyPrefix}-${platform}-go-${versionSpec}-${fileHash}`;
```

File: dist/setup/index.js
Comment (erezrokah) [inline dist/setup/index.js:63047]:
```suggestion
    const cacheKeyPrefix = core.getInput('cache-key-prefix') || 'setup-go';
    const primaryKey = `${cacheKeyPrefix}-${platform}-go-${versionSpec}-${fileHash}`;
```

Comment (erezrokah) [inline dist/setup/index.js:63047]:
```suggestion
    const cacheKeyPrefix = core.getInput('cache-key-prefix') || 'setup-go';
    const primaryKey = `${cacheKeyPrefix}-${platform}-go-${versionSpec}-${fileHash}`;
```

**Comment by erezrokah** [comment]
> @candiduslynx @erezrokah thanks for you note, i missed to click "submit" indeed. No the review is published

Thanks @dsame, I updated the PR and requested another review

**Comment by dsame** [comment]
@erezrokah can you please run `npm run format` and commit the format fixes? It is necessary to make these checks to pass: https://github.com/actions/setup-go/actions/runs/5155716632/jobs/9286746499?pr=366

**Comment by erezrokah** [comment]
> @erezrokah can you please run `npm run format` and commit the format fixes? It is necessary to make these checks to pass: https://github.com/actions/setup-go/actions/runs/5155716632/jobs/9286746499?pr=366

Woops, sorry about that. Fixed in https://github.com/actions/setup-go/pull/366/commits/ab4b1f7578bbd50ef75f9ad9a0a30a4bf707ad11

**Review by dsame — APPROVED** [review]

**Comment by andig** [comment]
Would be great to get a point release once this great PR is in to get cache keys plus tzst compression :) Thank you all for your work!

**Review by candiduslynx — APPROVED** [review]

**Review by marko-zivic-93 — APPROVED** [review]

**Comment by erezrokah** [comment]
Hi 👋 Anything blocking this from getting merged?

**Comment by andig** [comment]
Friendly ping- seems PR is stuck once more?

**Comment by andig** [comment]
PR is done and has three approvals :)

**Comment by dsame** [comment]
@erezrokah @andig while the PR is perfect by there's a hesitation about should it be merged because it repeats functionality of https://github.com/actions/cache action. Please let us a bit more time for discussing.

**Comment by andig** [comment]
@dsame Thank you for taking the time to carefully review!
I feel it does not repeat actions/cache functionality, but rather exposes it. If easier/clearer, you might potentially consider exposing cache-key directly instead of prefix. In any case, I think having to configure an additional key or prefix is much simpler from user perspective than setting up cache with correct folders, even if fully documented.

**Comment by dmitry-shibanov** [comment]
Hello @erezrokah. Sorry for the late response. You can prepare a note to the documentation.

**Comment by erezrokah** [comment]
Closing this PR. I'll open a new one for the docs

**Comment by andig** [comment]
I'm really sad this decision had to be made:

> In any case, I think having to configure an additional key or prefix is much simpler from user perspective than setting up cache with correct folders, even if fully documented.

We're trading a 2-line zero risk diff against setup complexity for every single user with more than the basic use case. That is imho a bad, high-cost trade-off.

**Pull request closed without merging**
